File: articles/computational-physics-and-scientific-simulation/c/diffusion_table.c
/*
Diffusion Table

This C workflow solves a one-dimensional diffusion equation using an
explicit finite-difference method.
*/

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    const double length_m = 1.0;
    const int n_grid = 201;
    const double diffusivity = 0.001;
    const double dt = 0.0002;
    const int n_steps = 1200;

    const double dx = length_m / (double)(n_grid - 1);
    const double s = diffusivity * dt / (dx * dx);

    if (s > 0.5) {
        fprintf(stderr, "Explicit diffusion stability condition violated.\n");
        return 1;
    }

    double *field = calloc(n_grid, sizeof(double));
    double *next_field = calloc(n_grid, sizeof(double));

    if (!field || !next_field) {
        fprintf(stderr, "Memory allocation failed.\n");
        return 1;
    }

    for (int i = 0; i < n_grid; i++) {
        double x = i * dx;
        field[i] = exp(-pow(x - 0.5, 2.0) / (2.0 * pow(0.04, 2.0)));
    }

    printf("step,time_s,total_integral,max_field,stability_number\n");

    for (int step = 0; step <= n_steps; step++) {
        if (step % 300 == 0) {
            double total = 0.0;
            double max_value = field[0];

            for (int i = 0; i < n_grid; i++) {
                total += field[i] * dx;
                if (field[i] > max_value) {
                    max_value = field[i];
                }
            }

            printf("%d,%.8f,%.12f,%.12f,%.8f\n",
                   step,
                   step * dt,
                   total,
                   max_value,
                   s);
        }

        for (int i = 0; i < n_grid; i++) {
            next_field[i] = field[i];
        }

        for (int i = 1; i < n_grid - 1; i++) {
            next_field[i] = field[i] + s * (field[i + 1] - 2.0 * field[i] + field[i - 1]);
        }

        next_field[0] = 0.0;
        next_field[n_grid - 1] = 0.0;

        double *temp = field;
        field = next_field;
        next_field = temp;
    }

    free(field);
    free(next_field);

    return 0;
}

File: articles/computational-physics-and-scientific-simulation/fortran/diffusion_table.f90
! Diffusion Table
!
! This Fortran workflow solves a one-dimensional diffusion equation using an
! explicit finite-difference method.

program diffusion_table
  implicit none

  integer, parameter :: n_grid = 201
  integer, parameter :: n_steps = 1200
  integer :: i, step_index
  real(8), parameter :: length_m = 1.0d0
  real(8), parameter :: diffusivity = 0.001d0
  real(8), parameter :: dt = 0.0002d0
  real(8) :: dx
  real(8) :: s
  real(8) :: x
  real(8), dimension(n_grid) :: field
  real(8), dimension(n_grid) :: next_field
  real(8) :: total
  real(8) :: max_field

  dx = length_m / real(n_grid - 1, 8)
  s = diffusivity * dt / (dx * dx)

  if (s > 0.5d0) then
     print *, "Explicit diffusion stability condition violated."
     stop
  end if

  do i = 1, n_grid
     x = real(i - 1, 8) * dx
     field(i) = exp(-((x - 0.5d0)**2) / (2.0d0 * 0.04d0**2))
  end do

  print *, "step,time_s,total_integral,max_field,stability_number"

  do step_index = 0, n_steps
     if (mod(step_index, 300) == 0) then
        total = sum(field) * dx
        max_field = maxval(field)

        print '(I8,A,F12.6,A,F16.10,A,F16.10,A,F12.8)', &
          step_index, ",", step_index * dt, ",", total, ",", max_field, ",", s
     end if

     next_field = field

     do i = 2, n_grid - 1
        next_field(i) = field(i) + s * (field(i + 1) - 2.0d0 * field(i) + field(i - 1))
     end do

     next_field(1) = 0.0d0
     next_field(n_grid) = 0.0d0

     field = next_field
  end do

end program diffusion_table
